labcontrol/gui/js_tests/js_tester.py

```diff
@@ -0,0 +1,130 @@
+import click
+from os.path import join, abspath, dirname
+from subprocess import Popen, PIPE
+import os
+
+
+'''
+Notes: It's generally not a good idea to have Click commands call other click
+commands, although ti can be done using context objects.
+
+This means that it's better to have an umbrella command where users can specify
+to turn off one or more tests, rather than have individual tests as separate
+commands, with an umbrella command to call them all.
+'''
+
+
+# used to establish the 'cli' group of commands.
+# all commands are specified on the command line as the first parameter.
+@click.group()
+def cli():
+    pass
+
+
+def run_js_test(timeout_in_milliseconds, html_file):
+    '''helper function accepts a Qunit html file referencing javascript tests,
+    executes it, and returns the results of the test.
+
+    Returns
+    -------
+    tuple (bool, str):  True if no errors occured, False otherwise.
+                        stdout from running tests.
+    '''
+
+    # params to node-qunit-puppeteer need to be encapsulated into one set of
+    # quotes
+    params = []
+    params.append('--allow-file-access-from-files')
+    params.append('--no-sandbox')
+    params = ' '.join(params)
+
+    cmd = []
+    cmd.append('node-qunit-puppeteer')
+    cmd.append(html_file)
+    cmd.append(str(timeout_in_milliseconds))
+
+    cmd.append('"%s"' % params)
+    cmd = ' '.join(cmd)
+
+    p = Popen(cmd, shell=True, stdout=PIPE, stderr=PIPE)
+    o, e = p.communicate()
+    o = o.decode("utf-8")
+    e = e.decode("utf-8")
+
+    # returncodes appear to be limited to 0 (success) and 1 (error).
+    # all information appears to write out stdout; stderr is empty
+    # on success or error.
+    if p.returncode == 0:
+        return(True, o)
+    else:
+        return(False, o)
+
+
+# Currently, there is only one command. Add more commands here.
+# note, the default has been explicitly defined here as 30,000ms. If none is
+# supplied, the node_qunit_puppeteer script itself will default to 30,000ms.
+@cli.command()
+@click.option('--timeout_in_milliseconds', required=False,
+              type=click.IntRange(0, None),
+              default=30000, show_default=True,
+              help='timeout for a unittest, in milliseconds')
+def all_tests(timeout_in_milliseconds):
+    '''Command to run all tests managed by this script. Currently there is only
+    one (javascript qunit tests).
```

Comment:
Guidance for 1-line docstring: "It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ..."." (PEP 257) ... can this be revised into this format?

Also, why "javascript qunit test"?  ... I would have expected this comment would reference "run_js_test"

Reply:
Will do!